### PR TITLE
WIP: Initial BIP157 support

### PR DESCRIPTION
**Pull request by torkelrogstad**
In this PR we: 

- Add basic support for BIP157 P2P messages 
- Introduce SPV mode and Neutrino mode to node
- Enhance logging and error handling in server main method 
- Improve error handling of logging configuration levels 
- Add trace logging of P2P message parsing 

To test this PR/do dev work on BIP157 you need to compile Jimpo's fork of Core, PR/branch is here: https://github.com/bitcoin/bitcoin/pull/16442/

TODO: 
- [x] Request compact filter headers if neutrino mode is enabled
- [x] Parse compact filter headers response
- [ ] Handle (store to disk, at least) received filter headers
- [ ] Add hard coded + symmetry tests to `getcfheaders` and `cfheaders` messages
- [ ] Copy documentation from BIP157 to new P2P messages

### CODE REVIEW

File: app/server/src/main/scala/org/bitcoins/server/ServerMain.scala [hunk 1]
```diff
@@ -40,17 +39,39 @@ object Main extends App {
...
   /** Log the given message, shut down the actor system and quit. */
   def error(message: Any): Nothing = {
-    logger.error(s"FATAL: $message")
-    logger.error(s"Shutting down actor system")
-    Await.result(system.terminate(), 10.seconds)
-    logger.error("Actor system terminated")
-    logger.error(s"Exiting")
-    sys.error(message.toString())
+    // we might want to log the stack trace if message is an error
+    val (processedMessage: Any, stackTrace: Option[Array[StackTraceElement]]) =
```
Comment (nkohen) [inline app/server/src/main/scala/org/bitcoins/server/ServerMain.scala:55]: nit: stackTraceOpt

File: app/server/src/main/scala/org/bitcoins/server/ServerMain.scala [hunk 2]
```diff
@@ -40,17 +39,39 @@ object Main extends App {
...
+      message match {
+        case maybeBoxed: ExecutionException =>
+          maybeBoxed.getCause() match {
+            case null                     => (maybeBoxed, None)
+            case err: NotImplementedError => (err, Some(err.getStackTrace()))
+            case other: Throwable         => (other, None)
```
Comment (nkohen) [inline app/server/src/main/scala/org/bitcoins/server/ServerMain.scala:61]: Do `other` and `maybeBoxed` in the `null` case not have a `getStackTrace` member?

File: core/src/main/scala/org/bitcoins/core/gcs/FilterType.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import org.bitcoins.core.protocol.NetworkElement
+import scodec.bits._
+import org.bitcoins.core.number.UInt64
+import org.bitcoins.core.number.UInt8
+import org.bitcoins.core.util.Factory
+
+/**
+  * Filter types for BIP157 block content filters
+  *
+  * @see [[https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#block-filters BIP158]]
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/gcs/FilterType.scala:12]: nit: I believe the comment should reference 158

File: core/src/main/scala/org/bitcoins/core/p2p/NetworkPayload.scala [hunk 4]
```diff
@@ -870,6 +873,155 @@ case object VerAckMessage extends ControlPayload {
...
+  def bytes: ByteVector = ???
+}
+
+object CompactFilterMessage extends Factory[CompactFilterMessage] {
+
+  /** Constructs a message from the tiven blockhash and filter */
+  def apply(
+      blockHash: DoubleSha256Digest,
+      filter: GolombFilter): CompactFilterMessage = {
+    val filterBytes = filter.bytes
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/p2p/NetworkPayload.scala:916]: Double-check that this doesn't already include the `CompactSizeUInt`, I forget whether or not it does

File: core/src/main/scala/org/bitcoins/core/p2p/NetworkPayload.scala [hunk 5]
```diff
@@ -870,6 +873,155 @@ case object VerAckMessage extends ControlPayload {
...
+  *
+  * TODO: doc on params
+  *
+  * @see [[https://github.com/bitcoin/bips/blob/master/bip-0157.mediawiki#cfheaders BIP157]]
+  */
+case class CompactFilterHeadersMessage(
+    filterType: FilterType,
+    stopHash: DoubleSha256Digest,
+    previousFilterHeader: DoubleSha256Digest,
+    filterHashes: Vector[DoubleSha256Digest])
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/p2p/NetworkPayload.scala:970]: You should create a more specific type `RawFilterHeader` that wraps a `DoubleSha256Digest` (and have `FilterHeader::hash` return one of these)

File: core/src/main/scala/org/bitcoins/core/p2p/NetworkPayload.scala [hunk 6]
```diff
@@ -870,6 +873,155 @@ case object VerAckMessage extends ControlPayload {
...
+  */
+case class GetCompactFilterHeadersMessage(
+    filterType: FilterType,
+    startHeight: UInt32,
+    stopHash: DoubleSha256Digest
+) extends DataPayload {
+  val commandName: String = NetworkPayload.getCompactFilterHeadersCommandName
+
+  def bytes: ByteVector =
+    RawGetcompactFilterHeadersMessageSerializer.write(this)
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/p2p/NetworkPayload.scala:937]: typo?

File: core/src/main/scala/org/bitcoins/core/p2p/ServiceIdentifier.scala [hunk 7]
```diff
@@ -82,7 +95,7 @@ sealed abstract class ServiceIdentifier extends NetworkElement {
     val innerText =
       if (nodeNone) "none"
       else
-        s"network=$nodeNetwork, getUtxo=$nodeGetUtxo, bloom=$nodeBloom, witness=$nodeWitness, xthin=$nodeXthin, networkLimited=$nodeNetworkLimited"
+        s"network=$nodeNetwork, compactFilters=$nodeCompactFilters getUtxo=$nodeGetUtxo, bloom=$nodeBloom, witness=$nodeWitness, xthin=$nodeXthin, networkLimited=$nodeNetworkLimited"
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/p2p/ServiceIdentifier.scala:98]: nit: missing comma

File: core/src/main/scala/org/bitcoins/core/p2p/ServiceIdentifier.scala [hunk 8]
```diff
@@ -143,6 +156,8 @@ object ServiceIdentifier extends Factory[ServiceIdentifier] {
     */
   val NODE_XTHIN: ServiceIdentifier = ServiceIdentifier(1 << 4)
 
+  val NODE_COMPACT_FILTERS = ServiceIdentifier(1 << 6)
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/p2p/ServiceIdentifier.scala:159]: doc

File: core/src/main/scala/org/bitcoins/core/serializers/p2p/messages/RawCompactFilterHeadersMessageSerializer.scala [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+      if (bytes.isEmpty) (hashes, ByteVector.empty)
+      else {
+        val (hashBytes, remainingBytes) = bytes.splitAt(32)
+        loop(hashes :+ DoubleSha256Digest.fromBytes(hashBytes), remainingBytes)
+      }
+
+    val (hashes, remainingBytes) = loop(Vector.empty, filterHashesBytes)
+
+    assert(hashes.length == filterHashesLength.toInt)
+    assert(remainingBytes.isEmpty)
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/serializers/p2p/messages/RawCompactFilterHeadersMessageSerializer.scala:44]: nit: Is it nicer to use `require` with error messages?

File: core/src/main/scala/org/bitcoins/core/serializers/p2p/messages/RawGetcompactFilterHeadersMessageSerializer.scala [hunk 10]
```diff
@@ -0,0 +1,13 @@
+package org.bitcoins.core.serializers.p2p.messages
+
+import org.bitcoins.core.serializers.RawBitcoinSerializer
+import org.bitcoins.core.p2p.GetCompactFilterHeadersMessage
+import scodec.bits.ByteVector
+
+object RawGetcompactFilterHeadersMessageSerializer
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/serializers/p2p/messages/RawGetcompactFilterHeadersMessageSerializer.scala:7]: typo

File: node/src/main/scala/org/bitcoins/node/networking/peer/DataMessageHandler.scala [hunk 11]
```diff
@@ -37,6 +46,24 @@ class DataMessageHandler(callbacks: SpvNodeCallbacks, chainHandler: ChainApi)(
       peerMsgSender: PeerMessageSender): Future[Unit] = {
 
     payload match {
+      case checkpoint: CompactFilterCheckPointMessage =>
```
Comment (nkohen) [inline node/src/main/scala/org/bitcoins/node/networking/peer/DataMessageHandler.scala:49]: ping: TODO

**Comment by torkelrogstad** [comment]
This was closed in favor of #695

**Pull request closed without merging**